Mesh shaders - initial wgpu hal changes

inner-daemons opened the pull request:
**Connections**
#3018 

**Description**
Introduces basic mesh shading functionality to wgpu-hal

**Testing**
This change doesn't itself contain any tests. However, the functionality is tested and an example is provided on a separate branch of my fork.

<!--
Thanks for filing! The codeowners file will automatically request reviews from the appropriate teams.

After you get a review and have addressed any comments, please explicitly re-request a review from the
person(s) who reviewed your changes. This will make sure it gets re-added to their review queue - you're no bothering us!
-->

**Checklist**

- [x] Run `cargo fmt`.
- [x] Run `taplo format`.
- [x] Run `cargo clippy`. If applicable, add:
  - [x] `--target wasm32-unknown-unknown`
- [x] Run `cargo xtask test` to run tests.
- [x] Add change to `CHANGELOG.md`. See simple instructions inside file.

**Comments**
This is one of my first PRs, so I probably did something wrong with regard to git or some other convention. Also, this is a very early draft PR. I just want some initial feedback.

inner-daemons committed:
Initial(untested commit), vulkan and gles only supported

inner-daemons committed:
Maybe fixed compiles for metal and dx12

inner-daemons committed:
Merge branch 'gfx-rs:trunk' into mesh-shading/wgpu-hal

inner-daemons committed:
Hopefully fixed compiles for other backends and updated to functional(?) vulkan thing

inner-daemons committed:
I don't get git
 Please enter a commit message to explain why this merge is necessary,

inner-daemons committed:
Fixed the clippy warning

inner-daemons committed:
Merge recent changes, most significantly features type overhaul

inner-daemons committed:
Fixed silly documentation mistake

cwfitzgerald commented:
Going to look at your proposal in a minute, but there's nothing really controversial in here, seems pretty straight forward. The spicy part is validation :)

inner-daemons commented:
It is saying it's failing some benchmarks in the CI, is that a concern? Otherwise, should I mark this as a non-draft PR? Thanks!

cwfitzgerald commented:
> It is saying it's failing some benchmarks in the CI, is that a concern

Yeah the error message is about a violation of a wgpu-hal precondition about features being violated, so there's something going on with wgpu-hal's features.

After that is fixed, feel free to mark as ready-to-go

inner-daemons committed:
Fixed issue with multiview feature

inner-daemons commented on wgpu-hal/src/vulkan/mod.rs:
```diff
@@ -460,7 +460,7 @@ pub struct Adapter {
     //queue_families: Vec<vk::QueueFamilyProperties>,
     known_memory_flags: vk::MemoryPropertyFlags,
     phd_capabilities: adapter::PhysicalDeviceProperties,
-    //phd_features: adapter::PhysicalDeviceFeatures,
```
Not sure why this was commented in the first place. Shouldn't be a big deal to make this actually available.

inner-daemons commented on wgpu-hal/src/vulkan/adapter.rs:
```diff
@@ -496,7 +505,11 @@ impl PhysicalDeviceFeatures {
+                        .multiview_mesh_shader(
+                            needed
+                                && multiview
+                                && phd_features.mesh_shader.unwrap().multiview_mesh_shader == 1,
```
Not sure if this is the best way to do this. I had to change stuff around so that whether this was supported was visible in this context. Worst comes to worst, we can forget multiview in mesh shaders for now. The main problem is that apparently llvmpipe supports mesh shaders and multiview but not multiview in mesh shaders.

cwfitzgerald replied:
Lets just forget it, and we'll tackle it as a separate feature.

inner-daemons replied:
Removed this

inner-daemons commented on wgpu-hal/src/vulkan/adapter.rs:
```diff
@@ -143,6 +147,9 @@ impl PhysicalDeviceFeatures {
         if let Some(ref mut feature) = self.robustness2 {
             info = info.push_next(feature);
         }
+        if let Some(ref mut feature) = self.multiview {
```
This was missing for some reason, I spent quite a lot of time very confused.

cwfitzgerald replied:
Do we just not have any tests for multiview....

Couldjafileanissue?

inner-daemons replied:
Done!

inner-daemons committed:
Dummy commit for dummy CI

The CI pooped itself, hopefully this fixes that. Will probably be undone either way.

inner-daemons commented:
@cwfitzgerald Sorry to bug you again. CI says there are 30 checks which haven’t completed, but are expected and waiting to be reported. But then it also says there are no checks to be done. Do I have to do something to kickstart the checking process? For some of the previous commits, it hasn’t needed any prompting. Otherwise, I think this PR is ready to be merged :)

cwfitzgerald commented:
> Do I have to do something to kickstart the checking process?

CI runs on the result of merging with trunk, so if there are conflicts with the merge, it can't run CI - so if you fix those conflicts, the CI will run.

cwfitzgerald commented:
Alright, will add this to my list to review properly before we land!

inner-daemons committed:
Merge branch 'trunk' into mesh-shading/wgpu-hal

inner-daemons committed:
Merge branch 'trunk' into mesh-shading/wgpu-hal

inner-daemons committed:
Re trigger CI checks, to avoid #7126

inner-daemons committed:
Merge branch 'trunk' into mesh-shading/wgpu-hal

inner-daemons committed:
Merge branch 'trunk' into mesh-shading/wgpu-hal

inner-daemons committed:
Merge branch 'trunk' into mesh-shading/wgpu-hal

inner-daemons committed:
Merge branch 'trunk' into mesh-shading/wgpu-hal

cwfitzgerald commented on wgpu-hal/src/dx12/command.rs:
```diff
@@ -1177,6 +1177,14 @@ impl crate::CommandEncoder for super::CommandEncoder {
             )
         }
     }
+    unsafe fn draw_mesh_tasks(
+        &mut self,
+        _group_count_x: u32,
+        _group_count_y: u32,
+        _group_count_z: u32,
+    ) {
+        unreachable!()
+    }
```
Could we file an issue about adding mesh shading to DX12, so we can add it to a mesh shader tracking issue

inner-daemons replied:
I'll get to that in a little bit

cwfitzgerald replied:
Poke :)

cwfitzgerald commented on wgpu-hal/src/lib.rs:
```diff
@@ -2145,6 +2175,34 @@ pub struct RenderPipelineDescriptor<
     /// The cache which will be used and filled when compiling this pipeline
     pub cache: Option<&'a Pc>,
 }
+// TODO: redesign this struct
```
Either:
- This todo should be fixed
- An issue should be filed and linked to here

Standalone todos tend to get lost in the goop

inner-daemons replied:
I think this struct is probably actually fine as is. I left this more as a note to myself to revisit, but it appears fine to me.

cwfitzgerald commented on wgpu-hal/src/vulkan/adapter.rs:
```diff
@@ -861,6 +903,9 @@ pub struct PhysicalDeviceProperties {
     /// `VK_EXT_robustness2` extension.
     robustness2: Option<vk::PhysicalDeviceRobustness2PropertiesEXT<'static>>,
 
+    /// This may be used at some point
```
I think a similar comment to the ones above would be fine here

inner-daemons replied:
Fixed!

cwfitzgerald commented on wgpu-hal/src/vulkan/command.rs:
```diff
@@ -1083,6 +1098,27 @@ impl crate::CommandEncoder for super::CommandEncoder {
+        offset: wgt::BufferAddress,
+        draw_count: u32,
+    ) {
+        match self.device.extension_fns.mesh_shading {
```
Nit: instead of doing a match, could this be a let-else, to keep the happy path in a straight line.

inner-daemons replied:
The reason I wrote it as a match is because that's how other draw commands handle it, for example see `draw_indirect_count`. I agree that an if-let would be preferable.

cwfitzgerald replied:
Lets go ahead and do it anyway, we can fix up the others later.

cwfitzgerald commented on wgpu-hal/src/vulkan/command.rs:
```diff
@@ -1135,6 +1171,34 @@ impl crate::CommandEncoder for super::CommandEncoder {
+        count_offset: wgt::BufferAddress,
+        max_count: u32,
+    ) {
+        if self.device.extension_fns.draw_indirect_count.is_none() {
+            panic!("Feature `DRAW_INDIRECT_COUNT` not enabled");
+        }
```
This feels weird, is this needed at all?

inner-daemons replied:
For the indirect count draw command, we do want to check for support I assume. Are you suggesting we check features instead of for the function pointers? Or do you have something else in mind?

cwfitzgerald replied:
> Provided by VK_EXT_mesh_shader with VK_KHR_draw_indirect_count or VK_VERSION_1_2

Ah I see

cwfitzgerald commented on wgpu-hal/src/vulkan/device.rs:
```diff
@@ -2115,6 +2115,240 @@ impl crate::Device for super::Device {
+
+        let compiled_ts = match desc.task_stage {
+            Some(ref stage) => {
+                // TODO: add proper naga stages
```
This should be an issue so we can track it

inner-daemons replied:
Agreed. I have another branch with only basic, non-functional naga changes that I plan to create a pull request from when this is finished and I have looked over it again.

cwfitzgerald commented on wgpu-hal/src/vulkan/device.rs:
```diff
@@ -2115,6 +2115,240 @@ impl crate::Device for super::Device {
+            unsafe { self.shared.set_object_name(raw, label) };
+        }
+
+        if let Some(CompiledStage {
+            temp_raw_module: Some(raw_module),
+            ..
+        }) = compiled_ts
+        {
+            unsafe { self.shared.raw.destroy_shader_module(raw_module, None) };
+        }
+        if let Some(raw_module) = compiled_ms.temp_raw_module {
+            unsafe { self.shared.raw.destroy_shader_module(raw_module, None) };
+        }
+        if let Some(CompiledStage {
+            temp_raw_module: Some(raw_module),
+            ..
+        }) = compiled_fs
+        {
+            unsafe { self.shared.raw.destroy_shader_module(raw_module, None) };
```
I assume you were vaguely copying from create_render_pipeline here, but this has the chance of leaking the shader modules if an earlier part of the function throws an error - don't worry about this for now, just wanted to note it.

inner-daemons replied:
I'll add a comment anyway.

cwfitzgerald commented on wgpu-types/src/features.rs:
```diff
@@ -1153,6 +1153,14 @@ bitflags_array! {
+
+        /// Enables mesh shaders and task shaders in mesh shader pipelines.
+        ///
+        /// Supported platforms:
+        /// - Vulkan (with [VK_EXT_mesh_shader](https://registry.khronos.org/vulkan/specs/latest/man/html/VK_EXT_mesh_shader.html))
```
Something I'm trying to do now is to list supported platforms and future platforms, indicating which platforms this _could_ support if the work is done.

inner-daemons replied:
Done!

cwfitzgerald reviewed (changes requested):
Got some comments, nothing show stopping.

inner-daemons committed:
Changes based on code review

inner-daemons committed:
Merge branch 'trunk' into mesh-shading/wgpu-hal

inner-daemons reviewed (commented):
I didn't mean to leave this review, oops

inner-daemons committed:
Fixed clippy warning, broken cargo.lock

inner-daemons committed:
Unfucked cargo.lock for real this time

inner-daemons committed:
Switched match to if-let in accordance with review

inner-daemons committed:
Updated changelog

inner-daemons committed:
Merge branch 'gfx-rs:trunk' into mesh-shading/wgpu-hal

inner-daemons committed:
Fix CI error

Done from web out of impatience

inner-daemons committed:
CI is very angry 😡

Made CI less angry by fixing formatting(hopefully). This commit was also done from GitHub web.

cwfitzgerald commented:
Going to re-review this on Monday

inner-daemons committed:
Merge branch 'trunk' into mesh-shading/wgpu-hal

inner-daemons committed:
Merge branch 'trunk' into mesh-shading/wgpu-hal

cwfitzgerald commented on wgpu-hal/src/vulkan/adapter.rs:
```diff
@@ -465,6 +482,32 @@ impl PhysicalDeviceFeatures {
+                        .mesh_shader(needed)
+                        .task_shader(needed)
+                        .multiview_mesh_shader(false),
+                    /*
+                    let multiview = requested_features.contains(wgt::Features::MULTIVIEW);
+
+                    .multiview_mesh_shader(
+                        needed
+                            && multiview
+                            && phd_features.mesh_shader.unwrap().multiview_mesh_shader == 1,
+                    )
+                    */
```
Let's fully remove this code, but we can drop a note about multiview_mesh_shader that points to the issue for dealing with multiview mesh shaders

cwfitzgerald reviewed (changes requested):
One teeny comment, then we're golden!

inner-daemons committed:
Removed comment in following request

inner-daemons commented:
> One teeny comment, then we're golden!

Alright, let’s get this landed!

cwfitzgerald commented on wgpu-hal/src/vulkan/adapter.rs:
```diff
@@ -465,6 +482,23 @@ impl PhysicalDeviceFeatures {
+                    vk::PhysicalDeviceMeshShaderFeaturesEXT::default()
+                        .mesh_shader(needed)
+                        .task_shader(needed)
+                        .multiview_mesh_shader(false),
```
```suggestion
                        // Multiview needs some special work https://github.com/gfx-rs/wgpu/issues/7262
                        .multiview_mesh_shader(false),
```

Missed one thing :)

cwfitzgerald committed:
Update wgpu-hal/src/vulkan/adapter.rs

inner-daemons committed:
Merge branch 'trunk' into mesh-shading/wgpu-hal
